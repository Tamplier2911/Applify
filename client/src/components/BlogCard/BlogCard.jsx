// import "./BlogCard.scss";
import React from "react";
import PropTypes from "prop-types";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { deleteOneBlogpostStart } from "../../redux/blogs/blogs.actions";

// components
import Button from "../Button/Button";

// data formaters
import getImageRelativePath from "../../utils/PathTransformations/getImageRelativePath";
import transformDateToLocaleString from "../../utils/DataTransformations/transformDateToLocaleString";

// JS Rendering CSS
import {
  BlogCardContainer,
  BlogCardHeader,
  BlogCardHeaderTitle,
  BlogCardHeaderTheme,
  BlogCardHeaderDate,
  BlogCardBody,
  BlogCardBodyImgWrap,
  BlogCardBodyImg,
  BlogCardBodyName,
  BlogCardBodyLikes,
  BlogCardBodyAmount,
  BlogCardBodySvgWrap,
  BlogCardBodySVG,
  BlogCardControlls,
  BlogCardControllsLink,
} from "./BlogCardStyles";

// component constants
import blogCardData from "./BlogCardConstants";

const BlogCard = ({ data, lang, deleteOneBlogpostStart }) => {
  const { _id, image, likes, createdAt, title, theme, author } = data
    ? data
    : {};
  const { blogCardDelete, blogCardUpdate } = blogCardData[lang];

  const blogImg = getImageRelativePath(image ? image : "");
  const blogDate = transformDateToLocaleString(
    createdAt ? createdAt : new Date()
  );
  const authorImg = getImageRelativePath(author ? author.photo : "");

  return (
    <BlogCardContainer
      img={blogImg}
      animateIn={"flipInY"}
      animateOnce={true}
      data-test="blog-card"
    >
      <BlogCardHeader data-test="blog-card-header">
        <BlogCardHeaderTitle>{title}</BlogCardHeaderTitle>
        <BlogCardHeaderTheme>{theme}</BlogCardHeaderTheme>
        <BlogCardHeaderDate>{blogDate}</BlogCardHeaderDate>
      </BlogCardHeader>
      <BlogCardBody data-test="blog-card-body">
        <BlogCardBodyImgWrap>
          <BlogCardBodyImg alt="happy author" src={authorImg} />
        </BlogCardBodyImgWrap>
        <BlogCardBodyName>{author ? author.name : ""}</BlogCardBodyName>
        <BlogCardBodyLikes>
          <BlogCardBodyAmount>{likes}</BlogCardBodyAmount>
          <BlogCardBodySvgWrap>
            <BlogCardBodySVG />
          </BlogCardBodySvgWrap>
        </BlogCardBodyLikes>
      </BlogCardBody>
      <BlogCardControlls data-test="blog-card-controls">
        <Button
          type="button"
          value={blogCardDelete}
          click={() => deleteOneBlogpostStart(_id)}
          data-test="blog-card-button"
        />
        <BlogCardControllsLink
          to={`/profile/blogs/${_id}`}
          data-test="blog-card-link"
        >
          {blogCardUpdate}
        </BlogCardControllsLink>
      </BlogCardControlls>
    </BlogCardContainer>
  );
};

BlogCard.propTypes = {
  data: PropTypes.shape({
    author: PropTypes.object,
    content: PropTypes.string,
    createdAt: PropTypes.string,
    image: PropTypes.string,
    likes: PropTypes.number,
    theme: PropTypes.string,
    title: PropTypes.string,
    _id: PropTypes.string,
  }),
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
});

export default connect(mapStateToProps, { deleteOneBlogpostStart })(BlogCard);
