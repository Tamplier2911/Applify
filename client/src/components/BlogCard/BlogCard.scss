/*

.blog-card {
  display: grid;
  grid-template-rows: 2.5fr 2fr 1fr;
  font-family: var(--font-title);
  grid-row-gap: 1rem;
  padding: 1rem;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);
  transition: box-shadow 0.3s;

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }

  &__header {
    display: grid;
    grid-auto-rows: min-content;
    // color: var(--cl-white);
    // padding: 1rem;
    // background-position: center;
    // background-size: cover;
    // background-repeat: no-repeat;
    text-align: center;

    background-color: var(--cl-white);
    padding: 1rem;

    &--title {
      font-size: 1.8rem;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }
    &--theme {
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }
    &--date {
    }
  }

  &__body {
    display: grid;
    position: relative;
    grid-auto-rows: min-content;
    align-content: end;

    background-color: var(--cl-white);
    padding: 1rem;

    &--imgwrap {
      position: absolute;
      left: 50%;
      top: -5rem;

      width: 9rem;
      height: 9rem;
      border-radius: 50%;
      overflow: hidden;
      box-shadow: 0rem 0rem 0.8rem var(--cl-primary);

      transform: translate(-50%);
    }

    &--img {
      width: 100%;
      height: 100%;
      display: block;
      object-fit: cover;
    }

    &--name {
      text-align: center;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }
    &--likes {
      display: grid;
      align-items: center;
      justify-items: center;
      grid-template-columns: repeat(2, 1fr);
    }
    &--amount {
    }
    &--svgwrap {
    }
    &--svg {
      width: 2rem;
      height: 2rem;
      fill: var(--cl-font);
    }
  }

  &__controlls {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-column-gap: 1rem;

    background-color: var(--cl-white);
    padding: 1rem;

    &--link {
      &:link,
      &:visited {
        display: grid;
        align-items: center;
        text-align: center;
        text-decoration: none;
        color: var(--cl-white);
        background-color: var(--cl-font);
        transition: background-color 0.3s;
      }

      &:hover,
      &:active {
        background-color: var(--cl-primary);
      }
    }
  }
}

*/
