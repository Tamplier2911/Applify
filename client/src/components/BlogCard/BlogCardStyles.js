import styled, { css } from "styled-components";
import { Link } from "react-router-dom";
import ScrollAnimation from "react-animate-on-scroll";

// svg
import { ReactComponent as HeartSVG } from "../../assets/svg/heart.svg";

const sharedStylesBg = css`
  background-color: var(--cl-white);
  padding: 1rem;
`;

const sharedStylesTextOverflow = css`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

const getBackgroundImage = (props) => {
  const { img } = props;
  return `background: center / cover no-repeat url(${img});`;
};

export const BlogCardContainer = styled(ScrollAnimation)`
  display: grid;
  grid-template-rows: 2.5fr 2fr 1fr;
  font-family: var(--font-title);
  grid-row-gap: 1rem;
  padding: 1rem;
  box-shadow: 0rem 0rem 0.8rem var(--cl-font);
  transition: box-shadow 0.3s;

  ${getBackgroundImage}

  &:hover {
    box-shadow: 0rem 0rem 0.8rem var(--cl-primary);
  }
`;

export const BlogCardHeader = styled.div`
  ${sharedStylesBg}
  display: grid;
  grid-auto-rows: min-content;
  text-align: center;
`;

export const BlogCardHeaderTitle = styled.div`
  ${sharedStylesTextOverflow}
  font-size: 1.8rem;
`;
export const BlogCardHeaderTheme = styled.div`
  ${sharedStylesTextOverflow}
`;

export const BlogCardHeaderDate = styled.div``;

export const BlogCardBody = styled.div`
  ${sharedStylesBg}
  display: grid;
  position: relative;
  grid-auto-rows: min-content;
  align-content: end;
`;
export const BlogCardBodyImgWrap = styled.div`
  position: absolute;
  left: 50%;
  top: -5rem;

  width: 9rem;
  height: 9rem;
  border-radius: 50%;
  overflow: hidden;
  box-shadow: 0rem 0rem 0.8rem var(--cl-primary);

  transform: translate(-50%);
`;

export const BlogCardBodyImg = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

export const BlogCardBodyName = styled.div`
  ${sharedStylesTextOverflow}
  text-align: center;
`;

export const BlogCardBodyLikes = styled.div`
  display: grid;
  align-items: center;
  justify-items: center;
  grid-template-columns: repeat(2, 1fr);
`;

export const BlogCardBodyAmount = styled.div``;

export const BlogCardBodySvgWrap = styled.div``;

export const BlogCardBodySVG = styled(HeartSVG)`
  width: 2rem;
  height: 2rem;
  fill: var(--cl-font);
`;

export const BlogCardControlls = styled.div`
  ${sharedStylesBg}
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-column-gap: 1rem;
`;

export const BlogCardControllsLink = styled(Link)`
  &:link,
  &:visited {
    display: grid;
    align-items: center;
    text-align: center;
    text-decoration: none;
    color: var(--cl-white);
    background-color: var(--cl-font);
    transition: background-color 0.3s;
  }

  &:hover,
  &:active {
    background-color: var(--cl-primary);
  }
`;
