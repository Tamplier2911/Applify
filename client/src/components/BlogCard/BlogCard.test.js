import React from "react";

// component
import BlogCard from "./BlogCard";

// utils
import {
  setShallowC,
  findByTestAttr,
  checkProps,
  storeFactory,
} from "../../tests/testUtils";

describe("<BlogCard />", () => {
  const initialStore = {
    lang: { languagePanelHidden: true, currentLanguage: "eng" },
  };

  const store = storeFactory(initialStore);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<BlogCard store={store} />);
  });

  it("renders without an error", () => {
    const blogCard = findByTestAttr(wrapper, "blog-card");
    expect(blogCard.length).toBe(1);
  });

  it("renders children - header", () => {
    const blogCardHeader = findByTestAttr(wrapper, "blog-card-header");
    expect(blogCardHeader.length).toBe(1);
  });

  it("renders children - body", () => {
    const blogCardBody = findByTestAttr(wrapper, "blog-card-body");
    expect(blogCardBody.length).toBe(1);
  });

  it("renders children - controlls", () => {
    const blogCardControls = findByTestAttr(wrapper, "blog-card-controls");
    expect(blogCardControls.length).toBe(1);
  });

  it("renders children - button", () => {
    const blogCardButton = findByTestAttr(wrapper, "blog-card-button");
    expect(blogCardButton.length).toBe(1);
  });

  it("renders children - link", () => {
    const blogCardLink = findByTestAttr(wrapper, "blog-card-link");
    expect(blogCardLink.length).toBe(1);
  });

  it("recieves expected prop types", () => {
    const expectedProps = {
      author: {},
      content: "str",
      createdAt: "str",
      image: "str",
      likes: 0,
      theme: "str",
      title: "str",
      _id: "str",
    };
    const err = checkProps(BlogCard, expectedProps, "data");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const blogCard = findByTestAttr(wrapper, "blog-card");
    expect(blogCard).toMatchSnapshot();
  });
});
