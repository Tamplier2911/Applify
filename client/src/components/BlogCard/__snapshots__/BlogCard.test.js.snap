// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogCard /> matches snapshot 1`] = `ShallowWrapper {}`;
